File: src/test/java/leetcode/LC024SwapNodesInPairs.java
package leetcode;

import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;

/**
 * https://leetcode.com/problems/swap-nodes-in-pairs/
 */
public class LC024SwapNodesInPairs {
    public class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }

    public class Solution {
        public ListNode swapPairs(ListNode head) {
            ListNode result = new ListNode(0);
            result.next = head;
            head = result;
            while (head.next != null && head.next.next != null) {
                ListNode first = head.next;
                ListNode second = head.next.next;
                ListNode after = head.next.next.next;
                head.next = second;
                second.next = first;
                first.next = after;
                head = head.next.next;
            }
            return result.next;
        }
    }

    @Test
    public void test_example() throws Exception {
        ListNode n1 = new ListNode(1);
        ListNode n2 = new ListNode(2);
        ListNode n3 = new ListNode(3);
        ListNode n4 = new ListNode(4);
        n1.next = n2;
        n2.next = n3;
        n3.next = n4;
        int[] expected = {2, 1, 4, 3};
        ListNode result = new Solution().swapPairs(n1);
        for (int value : expected) {
            assertNotEquals(null, result);
            assertEquals(value, result.val);
            result = result.next;
        }
    }
}

File: algorithms/code/leetcode/lc024_swap_nodes_in_pairs/lc024_swap_nodes_in_pairs.rs
// https://leetcode.com/problems/swap-nodes-in-pairs/
// #linking

// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

pub struct Solution;

impl Solution {
    pub fn swap_pairs(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut dummy = Some(Box::new(ListNode::new(0)));
        let mut current = &mut dummy;
        current.as_mut().unwrap().next = head;
        while let Some(mut first) = current.as_mut().unwrap().next.take() {
            if let Some(mut second) = first.next.take() {
                first.next = second.next.take();
                second.next = Some(first);
                current.as_mut().unwrap().next = Some(second);
                current = &mut current.as_mut().unwrap().next.as_mut().unwrap().next;
            } else {
                current.as_mut().unwrap().next = Some(first);
                break;
            }
        }
        dummy.unwrap().next
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn check(head: Option<Box<ListNode>>, expected: &[i32]) {
        let mut result = Solution::swap_pairs(head);
        for value in expected {
            assert_ne!(result, None);
            assert_eq!(result.as_ref().unwrap().val, *value);
            result = result.unwrap().next;
        }
    }

    #[test]
    fn test_example() {
        let mut n1 = Some(Box::new(ListNode::new(1)));
        let mut n2 = Some(Box::new(ListNode::new(2)));
        let mut n3 = Some(Box::new(ListNode::new(3)));
        let n4 = Some(Box::new(ListNode::new(4)));
        n3.as_mut().unwrap().next = n4;
        n2.as_mut().unwrap().next = n3;
        n1.as_mut().unwrap().next = n2;
        let expected = vec![2, 1, 4, 3];
        check(n1, &expected);
    }

    #[test]
    fn test_example_impaired() {
        let mut n1 = Some(Box::new(ListNode::new(1)));
        let mut n2 = Some(Box::new(ListNode::new(2)));
        let n3 = Some(Box::new(ListNode::new(3)));
        n2.as_mut().unwrap().next = n3;
        n1.as_mut().unwrap().next = n2;
        let expected = vec![2, 1, 3];
        check(n1, &expected);
    }

    #[test]
    fn test_nothing() {
        assert_eq!(Solution::swap_pairs(None), None);
    }
}
